Panic in Pool.Pick

Stack Trace:
```
[35m[2015-11-15T10:52:33.49Z] [GEARD] [geard.go:437] [CRITICAL] PANIC: runtime error: invalid memory address or nil pointer dereference
/usr/local/go/src/runtime/panic.go:423 (0x4321f9)
	gopanic: reflectcall(nil, unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz), uint32(d.siz))
/usr/local/go/src/runtime/panic.go:42 (0x4308b9)
	panicmem: panic(memoryError)
/usr/local/go/src/runtime/sigpanic_unix.go:24 (0x44770a)
	sigpanic: panicmem()
/go/src/github.com/channelmeter/gocql/connectionpool.go:188 (0x70456f)
	(*policyConnPool).Pick: conn = p.hostConnPools[host.Info().Peer].Pick(qry)
/go/src/github.com/channelmeter/gocql/session.go:429 (0x7463be)
	(*Session).executeBatch: host, conn := s.pool.Pick(nil)
/go/src/github.com/channelmeter/gocql/session.go:460 (0x746697)
	(*Session).ExecuteBatch: _, err := s.executeBatch(batch)
/go/src/github.com/channelmeter/geard/tasks/youtube/video.go:277 (0x5c3135)
```

I've been getting this during times of increased load, and when queries time out - also not sure what is nil in this situation `host`, `host.Info()` or `p.hostConnPools[host.Info().Peer]`.

I'm also at ad901edec6273959e31837afa3942b374efa2b5e, not sure if anything regarding this has changed since then.

Reopening because I still hit the issue under contention.

```
panic: policy *gocql.tokenAwareHostPolicy returned no host info: {info:<nil>}

goroutine 137 [running]:
github.com/channelmeter/gocql.(*policyConnPool).Pick(0xc2085c7b90, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/channelmeter/gocql/connectionpool.go:195 +0x2f8
github.com/channelmeter/gocql.(*controlConn).reconnect(0xc208589ec0, 0x7ff7840eb601)
	/go/src/github.com/channelmeter/gocql/control.go:86 +0x190
github.com/channelmeter/gocql.(*controlConn).heartBeat(0xc208589ec0)
	/go/src/github.com/channelmeter/gocql/control.go:57 +0x186
created by github.com/channelmeter/gocql.createControlConn
	/go/src/github.com/channelmeter/gocql/control.go:29 +0x172
```

What should be done in this situation? Here is the function:

```go
func (p *policyConnPool) Pick(qry *Query) (SelectedHost, *Conn) {
	nextHost := p.hostPolicy.Pick(qry)

	var (
		host SelectedHost
		conn *Conn
	)

	p.mu.RLock()
	defer p.mu.RUnlock()
	for conn == nil {
		host = nextHost()
		if host == nil {
			break
		} else if host.Info() == nil {
			panic(fmt.Sprintf("policy %T returned no host info: %+v", p.hostPolicy, host))
		}

		pool, ok := p.hostConnPools[host.Info().Peer]
		if !ok {
			continue
		}

		conn = pool.Pick(qry)
	}
	return host, conn
}
```

Return a nil host? Sleep and retry?

--- a/policies.go
+++ b/policies.go
@@ -94,7 +94,7 @@ func (r *roundRobinHostPolicy) SetPartitioner(partitioner string) {
 func (r *roundRobinHostPolicy) Pick(qry *Query) NextHost {
 	// i is used to limit the number of attempts to find a host
 	// to the number of hosts known to this policy
-	var i uint32
+	var i int
 	return func() SelectedHost {
 		r.mu.RLock()
 		defer r.mu.RUnlock()
@@ -102,14 +102,14 @@ func (r *roundRobinHostPolicy) Pick(qry *Query) NextHost {
 			return nil
 		}
 
-		var host *HostInfo
 		// always increment pos to evenly distribute traffic in case of
 		// failures
 		pos := atomic.AddUint32(&r.pos, 1)
-		if int(i) < len(r.hosts) {
-			host = &r.hosts[(pos)%uint32(len(r.hosts))]
-			i++
+		if i >= len(r.hosts) {
+			return nil
 		}
+		host := &r.hosts[(pos)%uint32(len(r.hosts))]
+		i++
 		return selectedRoundRobinHost{host}
 	}
 }
@@ -201,11 +201,9 @@ func (t *tokenAwareHostPolicy) Pick(qry *Query) NextHost {
 		return t.fallback.Pick(qry)
 	}
 
-	var host *HostInfo
-
 	t.mu.RLock()
 	// TODO retrieve a list of hosts based on the replication strategy
-	host = t.tokenRing.GetHostForPartitionKey(routingKey)
+	host := t.tokenRing.GetHostForPartitionKey(routingKey)
 	t.mu.RUnlock()
 
 	if host == nil {

--- a/policies_test.go
+++ b/policies_test.go
@@ -167,3 +167,28 @@ func TestRoundRobinConnPolicy(t *testing.T) {
 		t.Error("Expected conn1")
 	}
 }
+
+func TestRoundRobinNilHostInfo(t *testing.T) {
+	policy := RoundRobinHostPolicy()
+
+	host := HostInfo{HostId: "host-1"}
+	policy.SetHosts([]HostInfo{host})
+
+	iter := policy.Pick(nil)
+	next := iter()
+	if next == nil {
+		t.Fatal("got nil host")
+	} else if v := next.Info(); v == nil {
+		t.Fatal("got nil HostInfo")
+	} else if v.HostId != host.HostId {
+		t.Fatalf("expected host %v got %v", host, *v)
+	}
+
+	next = iter()
+	if next != nil {
+		t.Errorf("expected to get nil host got %+v", next)
+		if next.Info() == nil {
+			t.Fatalf("HostInfo is nil")
+		}
+	}
+}

